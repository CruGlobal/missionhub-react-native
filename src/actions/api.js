import lodashForEach from 'lodash/forEach';

import API_CALLS from '../api';
// import { logoutAction, toastAction } from './auth';
import apiRoutes from '../api/routes';
import { isObject } from '../utils/common';
import {
  EXPIRED_ACCESS_TOKEN,
  INVALID_ACCESS_TOKEN,
  INVALID_GRANT,
  UPDATE_TOKEN,
} from '../constants';

import { logout, refreshAccessToken, refreshAnonymousLogin } from './auth';
import { refreshMissionHubFacebookAccess } from './facebook';

// WARNING: You shouldn't have to touch this file to change routes/mapping
// Put new routes in '../api/routes';

// Setup the requests to be used for this file
let REQUESTS = {};
lodashForEach(apiRoutes, (data, key) => {
  REQUESTS[key] = {
    ...data,
    name: key,
    FETCH: `${key}_FETCH`,
    SUCCESS: `${key}_SUCCESS`,
  };
});
export { REQUESTS };

const METHODS_WITH_DATA = ['put', 'post', 'delete'];

export default function callApi(requestObject, query = {}, data = {}) {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      // Generic error handler
      const throwErr = msg => {
        if (__DEV__) {
          LOG(msg);
          throw new Error(msg);
        }
        reject(msg);
      };
      if (!requestObject) {
        return throwErr(
          `callApi(): There is no type: ${JSON.stringify(requestObject)}`,
        );
      }
      let newQuery = { ...query };
      const action = requestObject;

      const authState = getState().auth;
      if (!action.anonymous) {
        const { token } = authState;
        // If the request has not already passed in an access token, set it
        if (!newQuery.access_token) {
          newQuery.access_token = token;
        }
      }

      dispatch({
        query: newQuery,
        data: data || {},
        type: action.FETCH,
      });

      if (
        !API_CALLS[action.name] ||
        typeof API_CALLS[action.name] !== 'function'
      ) {
        return throwErr(
          `callApi(): API call is not a function: ${action.name}`,
        );
      }

      // If there is a method that uses data, call it here
      if (METHODS_WITH_DATA.includes(action.method)) {
        if (!action.anonymous && !newQuery.access_token) {
          return throwErr(
            `There is no token and route is not anonymous: ${JSON.stringify(
              action,
            )}`,
          );
        }
      }

      const handleError = err => {
        APILOG('REQUEST ERROR', action.name, err);
        const { apiError } = err;

        if (apiError) {
          if (
            apiError.errors &&
            apiError.errors[0] &&
            apiError.errors[0].detail
          ) {
            const errorDetail = apiError.errors[0].detail;
            if (
              errorDetail === EXPIRED_ACCESS_TOKEN ||
              errorDetail === INVALID_ACCESS_TOKEN
            ) {
              if (authState.refreshToken) {
                dispatch(refreshAccessToken());
              } else if (authState.isFirstTime) {
                dispatch(refreshAnonymousLogin());
              } else {
                dispatch(refreshMissionHubFacebookAccess());
              }
            }
          } else if (
            apiError.error === INVALID_GRANT &&
            action.name === REQUESTS.KEY_REFRESH_TOKEN.name
          ) {
            dispatch(logout(true));
          }
        }
        reject(err);
      };

      API_CALLS[action.name](newQuery, data)
        .then(response => {
          let actionResults = response ? response.results || {} : {};
          actionResults.response = response && response.response;
          let meta = response && response.meta;
          // If the results have an error object, call this to reject it
          if (
            (isObject(actionResults) && actionResults.error) ||
            actionResults.errors
          ) {
            handleError(actionResults);
            return;
          }

          // If there is a mapping function, call it
          if (action.mapResults) {
            actionResults = action.mapResults(
              actionResults,
              newQuery,
              data,
              getState,
            );
          }

          dispatch({
            results: actionResults || {},
            query: newQuery,
            data,
            meta,
            type: action.SUCCESS,
          });

          if (response.sessionHeader) {
            dispatch({
              type: UPDATE_TOKEN,
              token: response.sessionHeader,
            });
          }

          // Add data to the results to be used by followup actions
          actionResults.meta = meta;
          resolve(actionResults);
        })
        .catch(handleError);
    });
}
