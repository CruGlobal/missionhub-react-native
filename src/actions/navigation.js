import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
