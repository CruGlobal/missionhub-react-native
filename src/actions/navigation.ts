/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
