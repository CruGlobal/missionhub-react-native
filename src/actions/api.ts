/* eslint-disable @typescript-eslint/no-explicit-any, complexity */

import { ThunkAction } from 'redux-thunk';

import API_CALLS from '../api';
// import { logoutAction, toastAction } from './auth';
import { REQUESTS, ApiRouteConfigEntry } from '../api/routes';
import { isObject } from '../utils/common';
import {
  EXPIRED_ACCESS_TOKEN,
  INVALID_ACCESS_TOKEN,
  INVALID_GRANT,
  UPDATE_TOKEN,
} from '../constants';

import { logout } from './auth/auth';
import { refreshAccessToken } from './auth/key';
import { refreshAnonymousLogin } from './auth/anonymous';
import { refreshMissionHubFacebookAccess } from './auth/facebook';

// WARNING: You shouldn't have to touch this file to change routes/mapping
// Put new routes in '../api/routes';

const METHODS_WITH_DATA = ['put', 'post', 'delete'];

export default function callApi(
  action: ApiRouteConfigEntry,
  query: { [key: string]: string } = {},
  data: { [key: string]: any } = {},
): ThunkAction<
  void,
  any,
  null,
  | {
      query: any;
      data: any;
      type: string;
    }
  | {
      type: string;
      token: string;
    }
> {
  return async (dispatch, getState) => {
    // Generic error handler
    const throwErr = (msg: string) => {
      if (__DEV__) {
        // @ts-ignore
        LOG(msg);
        throw new Error(msg);
      }
      throw msg;
    };
    if (!action) {
      return throwErr(`callApi(): There is no type: ${JSON.stringify(action)}`);
    }
    const newQuery = {
      ...query,
    };

    const authState = getState().auth;
    if (!action.anonymous) {
      const { token } = authState;
      // If the request has not already passed in an access token, set it
      if (!newQuery.access_token) {
        newQuery.access_token = token;
      }
    }

    dispatch({
      query: newQuery,
      data: data || {},
      type: action.FETCH,
    });

    if (
      !API_CALLS[action.name] ||
      typeof API_CALLS[action.name] !== 'function'
    ) {
      return throwErr(`callApi(): API call is not a function: ${action.name}`);
    }

    // If there is a method that uses data, call it here
    if (action.method && METHODS_WITH_DATA.includes(action.method)) {
      if (
        !action.anonymous &&
        !newQuery.access_token &&
        !action.anonymousOptional
      ) {
        return throwErr(
          `There is no token and route is not anonymous: ${JSON.stringify({
            action,
            query: newQuery,
          })}`,
        );
      }
    }

    const handleError = (err: any) => {
      // @ts-ignore
      APILOG('REQUEST ERROR', action.name, err);
      const { apiError } = err;

      if (apiError) {
        if (
          apiError.errors &&
          apiError.errors[0] &&
          apiError.errors[0].detail
        ) {
          const errorDetail = apiError.errors[0].detail;
          if (
            errorDetail === EXPIRED_ACCESS_TOKEN ||
            errorDetail === INVALID_ACCESS_TOKEN
          ) {
            if (authState.refreshToken) {
              dispatch(refreshAccessToken());
            } else if (authState.isFirstTime) {
              dispatch(refreshAnonymousLogin());
            } else {
              dispatch(refreshMissionHubFacebookAccess());
            }
          }
        } else if (
          apiError.error === INVALID_GRANT &&
          action.name === REQUESTS.KEY_REFRESH_TOKEN.name
        ) {
          dispatch(logout(true));
        }
      }
      throw err;
    };

    try {
      const response = await API_CALLS[action.name](newQuery, data);
      const actionResults: { [key: string]: any } = response // TODO: replace any. I gave up typing this for now. It could be absolutely anything and every field is optional.
        ? response.results || {}
        : {};
      actionResults.response = response && response.response;
      const meta = response && response.meta;
      // If the results have an error object, call this to reject it
      if (
        (isObject(actionResults) && actionResults.error) ||
        actionResults.errors
      ) {
        handleError(actionResults);
        return;
      }

      dispatch({
        results: actionResults || {},
        query: newQuery,
        data,
        meta,
        type: action.SUCCESS,
      });

      if (response.sessionHeader) {
        dispatch({
          type: UPDATE_TOKEN,
          token: response.sessionHeader,
        });
      }

      // Add data to the results to be used by followup actions
      actionResults.meta = meta;
      return actionResults;
    } catch (error) {
      handleError(error);
    }
  };
}
