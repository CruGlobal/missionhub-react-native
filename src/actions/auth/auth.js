import {
  ACTIONS,
  CLEAR_UPGRADE_TOKEN,
  LOGOUT,
  MAIN_TABS,
} from '../../constants';
import { LANDING_SCREEN } from '../../containers/LandingScreen';
import { rollbar } from '../../utils/rollbar.config';
import { navigateReset } from '../navigation';
import { deletePushToken } from '../notifications';
import { trackActionWithoutData } from '../analytics';
import { SIGN_IN_FLOW } from '../../routes/constants';
import { GET_STARTED_SCREEN } from '../../containers/GetStartedScreen';
import { completeOnboarding } from '../onboardingProfile';
import { ADD_SOMEONE_SCREEN } from '../../containers/AddSomeoneScreen';

export function logout(forcedLogout = false) {
  return async dispatch => {
    await dispatch(deletePushToken());
    dispatch({ type: LOGOUT });
    dispatch(
      forcedLogout
        ? navigateReset(SIGN_IN_FLOW, { forcedLogout })
        : navigateReset(LANDING_SCREEN),
    );
    rollbar.clearPerson();
  };
}

export const retryIfInvalidatedClientToken = (
  firstAction,
  secondAction,
) => async dispatch => {
  // Historically we haven't cleared the client_token from redux after use,
  // so if the API throws a client_token invalidated error we retry this request
  // again without the client_token
  try {
    await dispatch(firstAction);
    dispatch({ type: CLEAR_UPGRADE_TOKEN });
  } catch (error) {
    const { apiError: { errors: [{ status, detail } = {}] = [] } = {} } = error;

    if (status === '422' && detail === 'client_token already invalidated') {
      await dispatch(secondAction);
      dispatch({ type: CLEAR_UPGRADE_TOKEN });
    } else {
      throw error;
    }
  }
};

export const navigateToPostAuthScreen = () => (dispatch, getState) => {
  const { person } = getState().auth;

  let nextScreen;

  if (!person.user.pathway_stage_id) {
    nextScreen = GET_STARTED_SCREEN;
    dispatch(trackActionWithoutData(ACTIONS.ONBOARDING_STARTED));
  } else if (hasPersonWithStageSelected(person)) {
    nextScreen = MAIN_TABS;
    dispatch(completeOnboarding());
  } else {
    nextScreen = ADD_SOMEONE_SCREEN;
    dispatch(trackActionWithoutData(ACTIONS.ONBOARDING_STARTED));
  }

  dispatch(navigateReset(nextScreen));
};

function hasPersonWithStageSelected(person) {
  return person.contact_assignments.some(contact => contact.pathway_stage_id);
}
