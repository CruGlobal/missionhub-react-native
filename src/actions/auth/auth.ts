import PushNotification from 'react-native-push-notification';
import { LoginManager } from 'react-native-fbsdk';
import { GoogleSignin } from '@react-native-community/google-signin';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';

import { LOGOUT } from '../../constants';
import { LANDING_SCREEN } from '../../containers/LandingScreen';
import { rollbar } from '../../utils/rollbar.config';
import { navigateReset } from '../navigation';
import { deletePushToken } from '../notifications';
import { SIGN_IN_FLOW } from '../../routes/constants';
import { apolloClient } from '../../apolloClient';
import { RootState } from '../../reducers';
import { deleteAllAuthTokens } from '../../auth/authStore';

export function logout(forcedLogout = false) {
  return async (dispatch: ThunkDispatch<RootState, never, AnyAction>) => {
    try {
      await dispatch(deletePushToken());
    } finally {
      dispatch(
        forcedLogout
          ? navigateReset(SIGN_IN_FLOW, { forcedLogout })
          : navigateReset(LANDING_SCREEN),
      );
      dispatch({ type: LOGOUT });
      apolloClient.clearStore();
      PushNotification.unregister();
      rollbar.clearPerson();
      try {
        LoginManager.logOut();
      } catch {}
      try {
        GoogleSignin.signOut();
      } catch {}
      await deleteAllAuthTokens();
    }
  };
}
