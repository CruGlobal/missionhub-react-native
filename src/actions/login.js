import { getPerson } from './people';
import { navigatePush } from './navigation';
import { updateLoggedInStatus } from './analytics';

export function onSuccessfulLogin() {
  return async(dispatch, getState) => {
    dispatch(updateLoggedInStatus(true));

    const personId = getState().auth.personId;
    const getMeResult = await dispatch(getPerson(personId));

    let nextScreen = 'GetStarted';
    if (getMeResult.findAll('user')[0].pathway_stage_id) {

      if (hasPersonWithStageSelected(getMeResult.find('person', personId))) {
        nextScreen = 'MainTabs';
      } else {
        nextScreen = 'AddSomeone';
      }
    }

    return dispatch(navigatePush(nextScreen));
  };
}

function hasPersonWithStageSelected(person) {
  return person.contact_assignments.some((contact) => contact.pathway_stage_id);
}