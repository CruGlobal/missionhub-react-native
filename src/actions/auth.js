import { THE_KEY_CLIENT_ID, LOGOUT, FIRST_TIME, ANALYTICS } from '../constants';
import { navigateReset, navigatePush } from './navigation';
import { getMe } from './people';
import { shouldRunSetUpPushNotifications, deletePushToken } from './notifications';
import { getStagesIfNotExists } from './stages';
import callApi, { REQUESTS } from './api';
import { updateAnalyticsContext } from './analytics';
import { onSuccessfulLogin } from './login';
import { LOGIN_SCREEN } from '../containers/LoginScreen';
import { LOGIN_OPTIONS_SCREEN } from '../containers/LoginOptionsScreen';

export function facebookLoginAction(accessToken, id, isUpgrade = false) {
  return (dispatch, getState) => {
    let data;
    const upgradeToken = getState().auth.token;
    if (isUpgrade) {
      data = {
        fb_access_token: accessToken,
        client_token: upgradeToken,
      };
    } else {
      data = {
        fb_access_token: accessToken,
      };
    }

    return dispatch(callApi(REQUESTS.FACEBOOK_LOGIN, {}, data)).then((results) => {
      LOG(results);
      dispatch(updateAnalyticsContext({ [ANALYTICS.FACEBOOK_ID]: id }));
      return dispatch(onSuccessfulLogin());
    });
  };
}

export function refreshAuth() {
  return async(dispatch, getState) => {
    const data = `grant_type=refresh_token&refresh_token=${getState().auth.refreshToken}`;

    await dispatch(callApi(REQUESTS.KEY_REFRESH_TOKEN, {}, data));
    dispatch(getKeyTicket());
  };
}

export function keyLogin(email, password) {
  const data = `grant_type=password&client_id=${THE_KEY_CLIENT_ID}&scope=fullticket%20extended&username=${email}&password=${password}`;

  return async(dispatch) => {
    await dispatch(callApi(REQUESTS.KEY_LOGIN, {}, data));
    await dispatch(getKeyTicket());

    return dispatch(onSuccessfulLogin());
  };
}

function getKeyTicket() {
  return async(dispatch) => {
    const keyTicketResult = await dispatch(callApi(REQUESTS.KEY_GET_TICKET, {}, {}));

    const data = { code: keyTicketResult.ticket };
    await dispatch(callApi(REQUESTS.TICKET_LOGIN, {}, data));
  };
}

export function codeLogin(code) {
  return (dispatch) => {
    return dispatch(callApi(REQUESTS.CREATE_MY_PERSON, {}, { code }))
      .then(() => {
        // Make sure this is set to FIRST_TIME so we know we're in the tryItNow flow
        dispatch(firstTime());
        return dispatch(onSuccessfulLogin());
      });
  };
}

export function logout() {
  return (dispatch, getState) => {
    const pushDeviceId = getState().notifications.pushDeviceId;
    if (!pushDeviceId) {
      dispatch(logoutReset());
    } else {
      dispatch(deletePushToken(pushDeviceId)).then(()=>{
        dispatch(logoutReset());
      }).catch(()=> {
        dispatch(logoutReset());
      });
    }
  };
}

export function logoutReset() {
  return (dispatch) => {
    dispatch({ type: LOGOUT });
    dispatch(navigateReset(LOGIN_SCREEN));
  };
}

export function upgradeAccount() {
  return (dispatch) => {
    dispatch(navigatePush(LOGIN_OPTIONS_SCREEN, { upgradeAccount: true }));
  };
}

export function firstTime() {
  return (dispatch) => {
    dispatch({ type: FIRST_TIME });
  };
}

// export function mergeAccount() {
//   return (dispatch) => {
//     return dispatch(callApi(REQUESTS.PROMOTE_ACCOUNT, {}, data))
//   };
// }

export function updateTimezone() {
  return (dispatch, getState) => {
    const currentTime = getState().auth.timezone;
    const timezone = new Date().getTimezoneOffset()/60*-1;
    if (currentTime !== `${timezone}`) {
      const data = {
        data: {
          attributes: {
            timezone: `${timezone}`,
          },
        },
      };
      return dispatch(callApi(REQUESTS.UPDATE_TIMEZONE, {}, data));
    }
  };
}

export function loadHome() {
  return (dispatch) => {
    // TODO: Set this up so it only loads these if it hasn't loaded them in X amount of time
    dispatch(getMe());
    dispatch(getStagesIfNotExists());
    dispatch(updateTimezone());
    dispatch(shouldRunSetUpPushNotifications());
  };
}
