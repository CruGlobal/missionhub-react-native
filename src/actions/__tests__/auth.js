/* eslint max-lines: 0, max-lines-per-function: 0 */

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import i18next from 'i18next';
import MockDate from 'mockdate';
import { Linking } from 'react-native';

import * as callApi from '../api';
import { REQUESTS } from '../api';
import * as constants from '../../constants';
import * as navigation from '../navigation';
import * as login from '../login';
import * as auth from '../auth';
import * as person from '../person';
import * as organizations from '../organizations';
import * as stages from '../stages';
import * as steps from '../steps';
import * as notifications from '../notifications';
import {
  keyLogin,
  refreshAccessToken,
  updateLocaleAndTimezone,
  codeLogin,
  logout,
  upgradeAccount,
  upgradeAccountSignIn,
  openKeyURL,
} from '../auth';
import { mockFnWithParams } from '../../../testUtils';
import { UPGRADE_ACCOUNT_SCREEN } from '../../containers/Auth/UpgradeAccountScreen';
import { OPEN_URL } from '../../constants';
import { getTimezoneString } from '../auth';
import { refreshAnonymousLogin } from '../auth';
import { deletePushToken } from '../notifications';
import * as onboardingProfile from '../onboardingProfile';
import { KEY_LOGIN_SCREEN } from '../../containers/Auth/KeyLoginScreen';

jest.mock('../../actions/notifications');

const email = 'klas&jflk@lkjasdf.com';
const password = 'this&is=unsafe';
const mockClientId = 123456;
const ticket = 'nfnvjvkfkfj886';
const data =
  `grant_type=password&client_id=${mockClientId}&scope=fullticket%20extended` +
  `&username=${encodeURIComponent(email)}&password=${encodeURIComponent(
    password,
  )}`;
const refreshToken = 'khjdsfkksadjhsladjjldsvajdscandjehrwewrqr';
const upgradeToken = '2d2123bd-8142-42e7-98e4-81a0dd7a87a6';
const mockStore = configureStore([thunk]);

let store;

constants.THE_KEY_CLIENT_ID = mockClientId;

const mockImplementation = implementation => {
  return jest.fn().mockImplementation(type => {
    return dispatch => {
      return dispatch(() => implementation(type));
    };
  });
};

const onSuccessfulLoginResult = { type: 'onSuccessfulLogin' };
const destinationAfterUpgrade = 'screen after upgrade';

beforeEach(() => {
  store = mockStore({
    auth: {
      token: 'testtoken',
      refreshToken,
      upgradeToken,
      person: {
        user: {},
      },
    },
  });

  mockFnWithParams(
    login,
    'onSuccessfulLogin',
    onSuccessfulLoginResult,
    destinationAfterUpgrade,
  );
});

describe('the key', () => {
  beforeEach(() => {
    callApi.default = mockImplementation(type => {
      if (type === REQUESTS.KEY_GET_TICKET) {
        return Promise.resolve({ ticket: ticket });
      } else {
        return Promise.resolve({});
      }
    });
  });

  describe('key refresh token', () => {
    it('should login to the key, then get a key ticket, then send the key ticket to Missionhub API, then handle successful login', () => {
      const refreshData = `grant_type=refresh_token&refresh_token=${refreshToken}`;

      return store.dispatch(refreshAccessToken()).then(() => {
        expect(callApi.default).toHaveBeenCalledWith(
          REQUESTS.KEY_REFRESH_TOKEN,
          {},
          refreshData,
        );
        expect(callApi.default).toHaveBeenCalledWith(
          REQUESTS.KEY_GET_TICKET,
          {},
          {},
        );
        expect(callApi.default).toHaveBeenCalledWith(
          REQUESTS.TICKET_LOGIN,
          {},
          { code: ticket },
        );
      });
    });
  });

  describe('key login', () => {
    it('should login to the key, then get a key ticket, then send the key ticket to Missionhub API, then handle successful login', () => {
      return store
        .dispatch(
          keyLogin(email, password, null, null, destinationAfterUpgrade),
        )
        .then(() => {
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.KEY_LOGIN,
            {},
            data,
          );
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.KEY_GET_TICKET,
            {},
            {},
          );
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.TICKET_LOGIN,
            {},
            { code: ticket },
          );

          expect(store.getActions()).toEqual([onSuccessfulLoginResult]);
        });
    });

    it('should login to the key, get a key ticket, then send the key ticket to Missionhub API with client token, then handle successful login', () => {
      return store
        .dispatch(
          keyLogin(email, password, null, true, destinationAfterUpgrade),
        )
        .then(() => {
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.KEY_LOGIN,
            {},
            data,
          );
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.KEY_GET_TICKET,
            {},
            {},
          );
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.TICKET_LOGIN,
            {},
            { code: ticket, client_token: upgradeToken },
          );

          expect(store.getActions()).toEqual([onSuccessfulLoginResult]);
        });
    });

    it('should send mfa code if passed', () => {
      const mfaCode = '123456';

      return store
        .dispatch(
          keyLogin(email, password, mfaCode, null, destinationAfterUpgrade),
        )
        .then(() => {
          expect(callApi.default).toHaveBeenCalledWith(
            REQUESTS.KEY_LOGIN,
            {},
            `${data}&thekey_mfa_token=${mfaCode}`,
          );

          expect(store.getActions()).toEqual([onSuccessfulLoginResult]);
        });
    });
  });

  describe('open key URL', () => {
    const expectedUrlResult = { type: OPEN_URL };

    it('should open key URL', () => {
      Linking.addEventListener = jest.fn((_, onComplete) => {
        onComplete({ url: 'testcode=result' });
      });
      Linking.removeEventListener = jest.fn();
      Linking.openURL = jest.fn();
      const onReturn = jest.fn();

      store.dispatch(openKeyURL('login?action=signup', onReturn, false));

      expect(Linking.addEventListener).toHaveBeenCalledWith(
        'url',
        expect.any(Function),
      );
      expect(Linking.removeEventListener).toHaveBeenCalledWith(
        'url',
        expect.any(Function),
      );
      expect(onReturn).toHaveBeenCalledTimes(1);
      expect(store.getActions()).toEqual([expectedUrlResult]);
      expect(Linking.openURL).toHaveBeenCalledWith(expect.any(String));
    });
  });
});

describe('code login', () => {
  beforeEach(() => {
    login.onSuccessfulLogin = jest.fn();
    auth.firstTime = jest.fn();
  });

  it('should run the code login and then on success', () => {
    store.dispatch(codeLogin('123')).then(() => {
      expect(login.onSuccessfulLogin).toHaveBeenCalledTimes(1);
    });
  });

  it('should run the code login and then first time', () => {
    store.dispatch(codeLogin('123')).then(() => {
      expect(auth.firstTime).toHaveBeenCalledTimes(1);
    });
  });
});

describe('updateLocaleAndTimezone', () => {
  beforeEach(() => {
    store = mockStore({
      auth: {
        person: {
          user: {
            timezone: '-8',
            language: 'fr-CA',
          },
        },
      },
    });
  });

  MockDate.set('2018-02-06', 300);
  i18next.language = 'en-US';

  const newUserSettings = {
    data: {
      attributes: {
        timezone: '-5',
        mobile_language: 'en-US',
      },
    },
  };

  it('should update timezone ', () => {
    store.dispatch(updateLocaleAndTimezone());
    expect(callApi.default).toHaveBeenCalledWith(
      REQUESTS.UPDATE_ME_USER,
      {},
      newUserSettings,
    );
  });
});

describe('refreshAnonymousLogin', () => {
  const apiResult = { type: 'refreshed anonymous token' };

  it('should send the code', async () => {
    mockFnWithParams(
      callApi,
      'default',
      dispatch => dispatch(apiResult),
      REQUESTS.REFRESH_ANONYMOUS_LOGIN,
      {},
      { code: upgradeToken },
    );

    await store.dispatch(refreshAnonymousLogin());

    expect(store.getActions()).toEqual([apiResult]);
  });
});

describe('logout', () => {
  it('should perform the needed actions for signing out', async () => {
    deletePushToken.mockReturnValue({
      type: REQUESTS.DELETE_PUSH_TOKEN.SUCCESS,
    });
    await store.dispatch(logout());
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should perform the needed actions for forced signing out', async () => {
    deletePushToken.mockReturnValue({
      type: REQUESTS.DELETE_PUSH_TOKEN.SUCCESS,
    });
    await store.dispatch(logout(true));
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('on upgrade account', () => {
  beforeEach(() => {
    navigation.navigatePush = jest.fn(screen => ({ type: screen }));
  });

  it('should navigate to login options page', async () => {
    const signupType = 'sign up';

    await store.dispatch(upgradeAccount(signupType));

    expect(navigation.navigatePush).toHaveBeenCalledWith(
      UPGRADE_ACCOUNT_SCREEN,
      {
        signupType,
      },
    );
    expect(store.getActions()).toEqual([{ type: UPGRADE_ACCOUNT_SCREEN }]);
  });

  it('should navigate to key login page', async () => {
    await store.dispatch(upgradeAccountSignIn());

    expect(store.getActions()).toEqual([{ type: KEY_LOGIN_SCREEN }]);
  });
});

describe('loadHome', () => {
  const getMeResult = { type: 'got me successfully' };
  const getStepsResult = { type: 'got steps successfully' };
  const getMyCommunitiesResult = { type: 'got communities' };
  const getStagesResult = { type: 'got stages' };
  const updateUserResult = { type: 'updated locale and TZ' };
  const notificationsResult = { type: 'show notification reminder' };
  const resetOnboardingPersonResult = { type: 'onboarding data cleared' };

  const userSettings = {
    data: {
      attributes: {
        timezone: getTimezoneString(),
        mobile_language: 'en-US',
      },
    },
  };

  it('loads me, organizations, stages, timezone, and notifications', async () => {
    mockFnWithParams(person, 'getMe', getMeResult);
    mockFnWithParams(steps, 'getMySteps', getStepsResult);
    mockFnWithParams(organizations, 'getMyCommunities', getMyCommunitiesResult);
    mockFnWithParams(stages, 'getStagesIfNotExists', getStagesResult);
    mockFnWithParams(
      callApi,
      'default',
      updateUserResult,
      REQUESTS.UPDATE_ME_USER,
      {},
      userSettings,
    );
    mockFnWithParams(notifications, 'showReminderOnLoad', notificationsResult);
    mockFnWithParams(
      onboardingProfile,
      'resetPerson',
      resetOnboardingPersonResult,
    );

    await store.dispatch(auth.loadHome());

    expect(store.getActions()).toEqual([
      getMeResult,
      getMyCommunitiesResult,
      getStagesResult,
      updateUserResult,
      resetOnboardingPersonResult,
      getStepsResult,
      notificationsResult,
    ]);
  });

  it('loads nothing because there is no token', async () => {
    store = mockStore({
      auth: {
        token: '',
        refreshToken,
        upgradeToken,
        person: {
          user: {},
        },
      },
    });
    await store.dispatch(auth.loadHome());

    expect(store.getActions()).toEqual([]);
  });
});
