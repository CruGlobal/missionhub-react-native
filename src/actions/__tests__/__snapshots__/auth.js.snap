// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logout should perform the needed actions for forced signing out 1`] = `
Array [
  Object {
    "type": "DELETE_PUSH_TOKEN_SUCCESS",
  },
  Object {
    "type": "app/LOGOUT",
  },
  Object {
    "actions": Array [
      Object {
        "params": Object {
          "forcedLogout": true,
        },
        "routeName": "nav/KEY_LOGIN",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "type": "Navigation/RESET",
  },
]
`;

exports[`logout should perform the needed actions for signing out 1`] = `
Array [
  Object {
    "type": "DELETE_PUSH_TOKEN_SUCCESS",
  },
  Object {
    "type": "app/LOGOUT",
  },
  Object {
    "actions": Array [
      Object {
        "params": Object {},
        "routeName": "nav/LANDING",
        "type": "Navigation/NAVIGATE",
      },
    ],
    "index": 0,
    "type": "Navigation/RESET",
  },
]
`;
